### PR TITLE
Set tenant on records during resource transformation

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

This PR will make sure that upon deployment the tenant id is taken from the command and written on the events. 

## Related issues

Part 1 of #13320 
Part 2 will consist of actually storing the tenant in the state

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/DecisionRequirementsMetadataRecord.java [hunk 1]
```diff
@@ -37,6 +37,7 @@ public class DecisionRequirementsMetadataRecord extends UnifiedRecordValue
   private final BinaryProperty checksumProp = new BinaryProperty("checksum");
 
   private final BooleanProperty isDuplicateProp = new BooleanProperty("isDuplicate", false);
+  private final StringProperty tenantIdProp = new StringProperty("tenantId", "");
```
Comment (remcowesterhoud) [inline protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/DecisionRequirementsMetadataRecord.java:40]: ❓ Should the default be `<default>` instead?

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentTest.java [hunk 2]
```diff
@@ -536,6 +538,116 @@ public void shouldCreateDeploymentResourceOfExecutablePools() {
...
+  public void shouldCreateProcessesForTenants() {
+    // given
+    final String tenant = "tenant";
+    final String tenant2 = "tenant2";
+
+    // when
+    final var deployment =
+        ENGINE.deployment().withXmlResource(process).withTenantId(tenant).deploy();
+    final var deployment2 =
+        ENGINE.deployment().withXmlResource(process).withTenantId(tenant2).deploy();
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentTest.java:621]: 🔧 Let's show that these are considered non-duplicate by asserting that the deployment key is different and the process definition keys differ

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentMultiplePartitionsTest.java [hunk 3]
```diff
@@ -43,13 +42,13 @@ public final class CreateDeploymentMultiplePartitionsTest {
...
   public static final int PARTITION_ID = DEPLOYMENT_PARTITION;
   public static final int PARTITION_COUNT = 3;
-  @ClassRule public static final EngineRule ENGINE = EngineRule.multiplePartition(PARTITION_COUNT);
   private static final BpmnModelInstance PROCESS =
       Bpmn.createExecutableProcess(PROCESS_ID).startEvent().endEvent().done();
   private static final BpmnModelInstance PROCESS_2 =
       Bpmn.createExecutableProcess("process2").startEvent().endEvent().done();
   private static final String DMN_DECISION_TABLE = "/dmn/decision-table.dmn";
   private static final String DMN_DECISION_TABLE_V2 = "/dmn/decision-table_v2.dmn";
+  @Rule public final EngineRule engine = EngineRule.multiplePartition(PARTITION_COUNT);
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentMultiplePartitionsTest.java:51]: ❓ Why is this necessary? 

🔧 This slows down these tests significantly. Let's find a solution to reuse the engine between the different cases if possible.

**Review by korthout — APPROVED** [review]
Nice and clean 🏗️ 

LGTM 👍

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/DecisionRequirementsMetadataRecord.java (hunk 1 above)
Comment (korthout) [reply protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/DecisionRequirementsMetadataRecord.java:40]: This will depend on our approach to postponing the migration tasks.

🔧 We could extract a TenantIdPropertyFactory or so, and reuse this in all records so we can easily control the default in one spot

Comment (koevskinikola) [reply protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/deployment/DecisionRequirementsMetadataRecord.java:40]: I think it would be safer, for the sake of validation to leave it empty.

We have the following cases:
* Multi-tenancy disabled:
   * Users can leave that property empty in their gRPC requests. Internally, Zeebe maps all data to the `<default>` tenant ID.
* Multi-tenancy enabled:
   * Users must define the `tenantId` property in their gRPC requests. Zeebe should reject requests/commands that don't set this property.

In the future, we will expose an API/interface in the Gateway so users may replace the `IdentityInterceptor` with their own implementation. We **might** also allow users to replace the `<default>` tenant ID with something else. So it would be better if we don't hard-code it in the records.

**Review by koevskinikola — APPROVED** [review]
LGTM :rocket:

**Comment by remcowesterhoud** [comment]
bors merge
